<?xml version='1.0' encoding='UTF-8'?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html>
        <head>
          <title>GenotypeRDD - ADAM_2.10: Core 0.21.1-SNAPSHOT API - org.bdgenomics.adam.rdd.variant.GenotypeRDD</title>
          <meta name="description" content="GenotypeRDD - ADAM 2.10 Core 0.21.1 - SNAPSHOT API - org.bdgenomics.adam.rdd.variant.GenotypeRDD" />
          <meta name="keywords" content="GenotypeRDD ADAM 2.10 Core 0.21.1 SNAPSHOT API org.bdgenomics.adam.rdd.variant.GenotypeRDD" />
          <meta http-equiv="content-type" content="text/html; charset=UTF-8" />
          
      <link href="../../../../../lib/template.css" media="screen" type="text/css" rel="stylesheet" />
      <link href="../../../../../lib/diagrams.css" media="screen" type="text/css" rel="stylesheet" id="diagrams-css" />
      <script type="text/javascript">
         if(top === self) {
            var url = '../../../../../index.html';
            var hash = 'org.bdgenomics.adam.rdd.variant.GenotypeRDD';
            var anchor = window.location.hash;
            var anchor_opt = '';
            if (anchor.length >= 1)
              anchor_opt = '@' + anchor.substring(1);
            window.location.href = url + '#' + hash + anchor_opt;
         }
   	  </script>
    
        </head>
        <body class="type">
      <div id="definition">
        <img src="../../../../../lib/class_big.png" />
        <p id="owner"><a href="../../../../package.html" class="extype" name="org">org</a>.<a href="../../../package.html" class="extype" name="org.bdgenomics">bdgenomics</a>.<a href="../../package.html" class="extype" name="org.bdgenomics.adam">adam</a>.<a href="../package.html" class="extype" name="org.bdgenomics.adam.rdd">rdd</a>.<a href="package.html" class="extype" name="org.bdgenomics.adam.rdd.variant">variant</a></p>
        <h1>GenotypeRDD</h1>
      </div>

      <h4 id="signature" class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">case class</span>
      </span>
      <span class="symbol">
        <span class="name">GenotypeRDD</span><span class="params">(<span name="rdd">rdd: <span class="extype" name="org.apache.spark.rdd.RDD">RDD</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="sequences">sequences: <a href="../../models/SequenceDictionary.html" class="extype" name="org.bdgenomics.adam.models.SequenceDictionary">SequenceDictionary</a></span>, <span name="samples">samples: <span class="extype" name="scala.Seq">Seq</span>[<span class="extype" name="org.bdgenomics.formats.avro.Sample">Sample</span>]</span>, <span name="headerLines">headerLines: <span class="extype" name="scala.Seq">Seq</span>[<span class="extype" name="htsjdk.variant.vcf.VCFHeaderLine">VCFHeaderLine</span>] = <span class="defval" name="org.bdgenomics.adam.converters.DefaultHeaderLines.allHeaderLines">...</span></span>)</span><span class="result"> extends <a href="../MultisampleAvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.MultisampleAvroGenomicRDD">MultisampleAvroGenomicRDD</a>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>] with <span class="extype" name="scala.Product">Product</span> with <span class="extype" name="scala.Serializable">Serializable</span></span>
      </span>
      </h4>
      
          <div id="comment" class="fullcommenttop"><div class="comment cmt"><p>An RDD containing genotypes called in a set of samples against a given
reference genome.
</p></div><dl class="paramcmts block"><dt class="param">rdd</dt><dd class="cmt"><p>Called genotypes.</p></dd><dt class="param">sequences</dt><dd class="cmt"><p>A dictionary describing the reference genome.</p></dd><dt class="param">samples</dt><dd class="cmt"><p>The samples called.</p></dd><dt class="param">headerLines</dt><dd class="cmt"><p>The VCF header lines that cover all INFO/FORMAT fields
  needed to represent this RDD of Genotypes.
</p></dd></dl><div class="toggleContainer block">
          <span class="toggle">Linear Supertypes</span>
          <div class="superTypes hiddenContent"><span class="extype" name="scala.Product">Product</span>, <span class="extype" name="scala.Equals">Equals</span>, <a href="../MultisampleAvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.MultisampleAvroGenomicRDD">MultisampleAvroGenomicRDD</a>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>], <a href="../MultisampleGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.MultisampleGenomicRDD">MultisampleGenomicRDD</a>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>], <a href="../AvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.AvroGenomicRDD">AvroGenomicRDD</a>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>], <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>], <span class="extype" name="org.bdgenomics.adam.rdd.ADAMRDDFunctions">ADAMRDDFunctions</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="org.bdgenomics.utils.misc.Logging">Logging</span>, <span class="extype" name="scala.Serializable">Serializable</span>, <span class="extype" name="java.io.Serializable">Serializable</span>, <span class="extype" name="scala.AnyRef">AnyRef</span>, <span class="extype" name="scala.Any">Any</span></div>
        </div></div>
        

      <div id="mbrsel">
        <div id="textfilter"><span class="pre"></span><span class="input"><input id="mbrsel-input" type="text" accesskey="/" /></span><span class="post"></span></div>
        <div id="order">
              <span class="filtertype">Ordering</span>
              <ol>
                
                <li class="alpha in"><span>Alphabetic</span></li>
                <li class="inherit out"><span>By inheritance</span></li>
              </ol>
            </div>
        <div id="ancestors">
                <span class="filtertype">Inherited<br />
                </span>
                <ol id="linearization">
                  <li class="in" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD"><span>GenotypeRDD</span></li><li class="in" name="scala.Product"><span>Product</span></li><li class="in" name="scala.Equals"><span>Equals</span></li><li class="in" name="org.bdgenomics.adam.rdd.MultisampleAvroGenomicRDD"><span>MultisampleAvroGenomicRDD</span></li><li class="in" name="org.bdgenomics.adam.rdd.MultisampleGenomicRDD"><span>MultisampleGenomicRDD</span></li><li class="in" name="org.bdgenomics.adam.rdd.AvroGenomicRDD"><span>AvroGenomicRDD</span></li><li class="in" name="org.bdgenomics.adam.rdd.GenomicRDD"><span>GenomicRDD</span></li><li class="in" name="org.bdgenomics.adam.rdd.ADAMRDDFunctions"><span>ADAMRDDFunctions</span></li><li class="in" name="org.bdgenomics.utils.misc.Logging"><span>Logging</span></li><li class="in" name="scala.Serializable"><span>Serializable</span></li><li class="in" name="java.io.Serializable"><span>Serializable</span></li><li class="in" name="scala.AnyRef"><span>AnyRef</span></li><li class="in" name="scala.Any"><span>Any</span></li>
                </ol>
              </div><div id="ancestors">
            <span class="filtertype"></span>
            <ol>
              <li class="hideall out"><span>Hide All</span></li>
              <li class="showall in"><span>Show all</span></li>
            </ol>
            <a href="http://docs.scala-lang.org/overviews/scaladoc/usage.html#members" target="_blank">Learn more about member selection</a>
          </div>
        <div id="visbl">
            <span class="filtertype">Visibility</span>
            <ol><li class="public in"><span>Public</span></li><li class="all out"><span>All</span></li></ol>
          </div>
      </div>

      <div id="template">
        <div id="allMembers">
        <div id="constructors" class="members">
              <h3>Instance Constructors</h3>
              <ol><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#&lt;init&gt;" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="&lt;init&gt;(rdd:org.apache.spark.rdd.RDD[org.bdgenomics.formats.avro.Genotype],sequences:org.bdgenomics.adam.models.SequenceDictionary,samples:Seq[org.bdgenomics.formats.avro.Sample],headerLines:Seq[htsjdk.variant.vcf.VCFHeaderLine]):org.bdgenomics.adam.rdd.variant.GenotypeRDD"></a>
      <a id="&lt;init&gt;:GenotypeRDD"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">new</span>
      </span>
      <span class="symbol">
        <span class="name">GenotypeRDD</span><span class="params">(<span name="rdd">rdd: <span class="extype" name="org.apache.spark.rdd.RDD">RDD</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="sequences">sequences: <a href="../../models/SequenceDictionary.html" class="extype" name="org.bdgenomics.adam.models.SequenceDictionary">SequenceDictionary</a></span>, <span name="samples">samples: <span class="extype" name="scala.Seq">Seq</span>[<span class="extype" name="org.bdgenomics.formats.avro.Sample">Sample</span>]</span>, <span name="headerLines">headerLines: <span class="extype" name="scala.Seq">Seq</span>[<span class="extype" name="htsjdk.variant.vcf.VCFHeaderLine">VCFHeaderLine</span>] = <span class="defval" name="org.bdgenomics.adam.converters.DefaultHeaderLines.allHeaderLines">...</span></span>)</span>
      </span>
      </h4>
      <p class="shortcomment cmt"></p><div class="fullcomment"><div class="comment cmt"></div><dl class="paramcmts block"><dt class="param">rdd</dt><dd class="cmt"><p>Called genotypes.</p></dd><dt class="param">sequences</dt><dd class="cmt"><p>A dictionary describing the reference genome.</p></dd><dt class="param">samples</dt><dd class="cmt"><p>The samples called.</p></dd><dt class="param">headerLines</dt><dd class="cmt"><p>The VCF header lines that cover all INFO/FORMAT fields
  needed to represent this RDD of Genotypes.
</p></dd></dl></div>
    </li></ol>
            </div>

        

        

        <div id="values" class="values members">
              <h3>Value Members</h3>
              <ol><li name="scala.AnyRef#!=" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="!=(x$1:AnyRef):Boolean"></a>
      <a id="!=(AnyRef):Boolean"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span title="gt4s: $bang$eq" class="name">!=</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.AnyRef">AnyRef</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="scala.Any#!=" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="!=(x$1:Any):Boolean"></a>
      <a id="!=(Any):Boolean"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span title="gt4s: $bang$eq" class="name">!=</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.Any">Any</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>Any</dd></dl></div>
    </li><li name="scala.AnyRef###" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="##():Int"></a>
      <a id="##():Int"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span title="gt4s: $hash$hash" class="name">##</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Int">Int</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef → Any</dd></dl></div>
    </li><li name="scala.AnyRef#==" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="==(x$1:AnyRef):Boolean"></a>
      <a id="==(AnyRef):Boolean"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span title="gt4s: $eq$eq" class="name">==</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.AnyRef">AnyRef</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="scala.Any#==" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="==(x$1:Any):Boolean"></a>
      <a id="==(Any):Boolean"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span title="gt4s: $eq$eq" class="name">==</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.Any">Any</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>Any</dd></dl></div>
    </li><li name="scala.Any#asInstanceOf" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="asInstanceOf[T0]:T0"></a>
      <a id="asInstanceOf[T0]:T0"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">asInstanceOf</span><span class="tparams">[<span name="T0">T0</span>]</span><span class="result">: <span class="extype" name="scala.Any.asInstanceOf.T0">T0</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>Any</dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#broadcastRegionJoin" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="broadcastRegionJoin[X,Y&lt;:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],Z&lt;:org.bdgenomics.adam.rdd.GenomicRDD[(T,X),Z]](genomicRdd:org.bdgenomics.adam.rdd.GenomicRDD[X,Y])(implicittTag:scala.reflect.ClassTag[T],implicitxTag:scala.reflect.ClassTag[X]):org.bdgenomics.adam.rdd.GenomicRDD[(T,X),Z]"></a>
      <a id="broadcastRegionJoin[X,Y&lt;:GenomicRDD[X,Y],Z&lt;:GenomicRDD[(Genotype,X),Z]](GenomicRDD[X,Y])(ClassTag[Genotype],ClassTag[X]):GenomicRDD[(Genotype,X),Z]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">broadcastRegionJoin</span><span class="tparams">[<span name="X">X</span>, <span name="Y">Y &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoin.Y">Y</span>]</span>, <span name="Z">Z &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoin.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoin.Z">Z</span>]</span>]</span><span class="params">(<span name="genomicRdd">genomicRdd: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoin.Y">Y</span>]</span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tTag">tTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="xTag">xTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoin.X">X</span>]</span>)</span><span class="result">: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoin.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoin.Z">Z</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">Performs a broadcast inner join between this RDD and another RDD.</p><div class="fullcomment"><div class="comment cmt"><p>Performs a broadcast inner join between this RDD and another RDD.</p><p>In a broadcast join, the left RDD (this RDD) is collected to the driver,
and broadcast to all the nodes in the cluster. The key equality function
used for this join is the reference region overlap function. Since this
is an inner join, all values who do not overlap a value from the other
RDD are dropped.
</p></div><dl class="paramcmts block"><dt class="param">genomicRdd</dt><dd class="cmt"><p>The right RDD in the join.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new genomic RDD containing all pairs of keys that
  overlapped in the genomic coordinate space.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#broadcastRegionJoinAndGroupByRight" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="broadcastRegionJoinAndGroupByRight[X,Y&lt;:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],Z&lt;:org.bdgenomics.adam.rdd.GenomicRDD[(Iterable[T],X),Z]](genomicRdd:org.bdgenomics.adam.rdd.GenomicRDD[X,Y])(implicittTag:scala.reflect.ClassTag[T],implicitxTag:scala.reflect.ClassTag[X]):org.bdgenomics.adam.rdd.GenomicRDD[(Iterable[T],X),Z]"></a>
      <a id="broadcastRegionJoinAndGroupByRight[X,Y&lt;:GenomicRDD[X,Y],Z&lt;:GenomicRDD[(Iterable[Genotype],X),Z]](GenomicRDD[X,Y])(ClassTag[Genotype],ClassTag[X]):GenomicRDD[(Iterable[Genotype],X),Z]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">broadcastRegionJoinAndGroupByRight</span><span class="tparams">[<span name="X">X</span>, <span name="Y">Y &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoinAndGroupByRight.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoinAndGroupByRight.Y">Y</span>]</span>, <span name="Z">Z &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Iterable">Iterable</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoinAndGroupByRight.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoinAndGroupByRight.Z">Z</span>]</span>]</span><span class="params">(<span name="genomicRdd">genomicRdd: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoinAndGroupByRight.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoinAndGroupByRight.Y">Y</span>]</span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tTag">tTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="xTag">xTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoinAndGroupByRight.X">X</span>]</span>)</span><span class="result">: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Iterable">Iterable</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoinAndGroupByRight.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.broadcastRegionJoinAndGroupByRight.Z">Z</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">Performs a broadcast inner join between this RDD and another RDD.</p><div class="fullcomment"><div class="comment cmt"><p>Performs a broadcast inner join between this RDD and another RDD.</p><p>In a broadcast join, the left RDD (this RDD) is collected to the driver,
and broadcast to all the nodes in the cluster. The key equality function
used for this join is the reference region overlap function. Since this
is an inner join, all values who do not overlap a value from the other
RDD are dropped.
</p></div><dl class="paramcmts block"><dt class="param">genomicRdd</dt><dd class="cmt"><p>The right RDD in the join.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new genomic RDD containing all pairs of keys that
  overlapped in the genomic coordinate space.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#buildTree" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="buildTree(rdd:org.apache.spark.rdd.RDD[(org.bdgenomics.adam.models.ReferenceRegion,org.bdgenomics.formats.avro.Genotype)])(implicittTag:scala.reflect.ClassTag[org.bdgenomics.formats.avro.Genotype]):org.bdgenomics.utils.interval.array.IntervalArray[org.bdgenomics.adam.models.ReferenceRegion,org.bdgenomics.formats.avro.Genotype]"></a>
      <a id="buildTree(RDD[(ReferenceRegion,Genotype)])(ClassTag[Genotype]):IntervalArray[ReferenceRegion,Genotype]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">buildTree</span><span class="params">(<span name="rdd">rdd: <span class="extype" name="org.apache.spark.rdd.RDD">RDD</span>[(<a href="../../models/ReferenceRegion.html" class="extype" name="org.bdgenomics.adam.models.ReferenceRegion">ReferenceRegion</a>, <span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>)]</span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tTag">tTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>)</span><span class="result">: <span class="extype" name="org.bdgenomics.utils.interval.array.IntervalArray">IntervalArray</span>[<a href="../../models/ReferenceRegion.html" class="extype" name="org.bdgenomics.adam.models.ReferenceRegion">ReferenceRegion</a>, <span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd><a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a> → <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="scala.AnyRef#clone" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="clone():Object"></a>
      <a id="clone():AnyRef"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">clone</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.AnyRef">AnyRef</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected[<a href="../../../../../java$lang.html" class="extype" name="java.lang">java.lang</a>] </dd><dt>Definition Classes</dt><dd>AnyRef</dd><dt>Annotations</dt><dd>
                <span class="name">@throws</span><span class="args">(<span>
      
      <span class="defval" name="classOf[java.lang.CloneNotSupportedException]">...</span>
    </span>)</span>
              
        </dd></dl></div>
    </li><li name="scala.AnyRef#eq" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="eq(x$1:AnyRef):Boolean"></a>
      <a id="eq(AnyRef):Boolean"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">eq</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.AnyRef">AnyRef</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#filterByOverlappingRegion" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="filterByOverlappingRegion(query:org.bdgenomics.adam.models.ReferenceRegion):U"></a>
      <a id="filterByOverlappingRegion(ReferenceRegion):GenotypeRDD"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">filterByOverlappingRegion</span><span class="params">(<span name="query">query: <a href="../../models/ReferenceRegion.html" class="extype" name="org.bdgenomics.adam.models.ReferenceRegion">ReferenceRegion</a></span>)</span><span class="result">: <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Runs a filter that selects data in the underlying RDD that overlaps a
single genomic region.</p><div class="fullcomment"><div class="comment cmt"><p>Runs a filter that selects data in the underlying RDD that overlaps a
single genomic region.
</p></div><dl class="paramcmts block"><dt class="param">query</dt><dd class="cmt"><p>The region to query for.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new GenomicRDD containing only data that overlaps the
  query region.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#filterByOverlappingRegions" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="filterByOverlappingRegions(querys:List[org.bdgenomics.adam.models.ReferenceRegion]):U"></a>
      <a id="filterByOverlappingRegions(List[ReferenceRegion]):GenotypeRDD"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">filterByOverlappingRegions</span><span class="params">(<span name="querys">querys: <span class="extype" name="scala.List">List</span>[<a href="../../models/ReferenceRegion.html" class="extype" name="org.bdgenomics.adam.models.ReferenceRegion">ReferenceRegion</a>]</span>)</span><span class="result">: <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Runs a filter that selects data in the underlying RDD that overlaps several genomic regions.</p><div class="fullcomment"><div class="comment cmt"><p>Runs a filter that selects data in the underlying RDD that overlaps several genomic regions.
</p></div><dl class="paramcmts block"><dt class="param">querys</dt><dd class="cmt"><p>The regions to query for.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new GenomicRDD containing only data that overlaps the
  querys region.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="scala.AnyRef#finalize" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="finalize():Unit"></a>
      <a id="finalize():Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">finalize</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected[<a href="../../../../../java$lang.html" class="extype" name="java.lang">java.lang</a>] </dd><dt>Definition Classes</dt><dd>AnyRef</dd><dt>Annotations</dt><dd>
                <span class="name">@throws</span><span class="args">(<span>
      
      <span class="symbol">classOf[java.lang.Throwable]</span>
    </span>)</span>
              
        </dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#flattenRddByRegions" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="flattenRddByRegions():org.apache.spark.rdd.RDD[(org.bdgenomics.adam.models.ReferenceRegion,T)]"></a>
      <a id="flattenRddByRegions():RDD[(ReferenceRegion,Genotype)]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">flattenRddByRegions</span><span class="params">()</span><span class="result">: <span class="extype" name="org.apache.spark.rdd.RDD">RDD</span>[(<a href="../../models/ReferenceRegion.html" class="extype" name="org.bdgenomics.adam.models.ReferenceRegion">ReferenceRegion</a>, <span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>)]</span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#fullOuterShuffleRegionJoin" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="fullOuterShuffleRegionJoin[X,Y&lt;:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],Z&lt;:org.bdgenomics.adam.rdd.GenomicRDD[(Option[T],Option[X]),Z]](genomicRdd:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],optPartitions:Option[Int])(implicittTag:scala.reflect.ClassTag[T],implicitxTag:scala.reflect.ClassTag[X]):org.bdgenomics.adam.rdd.GenomicRDD[(Option[T],Option[X]),Z]"></a>
      <a id="fullOuterShuffleRegionJoin[X,Y&lt;:GenomicRDD[X,Y],Z&lt;:GenomicRDD[(Option[Genotype],Option[X]),Z]](GenomicRDD[X,Y],Option[Int])(ClassTag[Genotype],ClassTag[X]):GenomicRDD[(Option[Genotype],Option[X]),Z]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">fullOuterShuffleRegionJoin</span><span class="tparams">[<span name="X">X</span>, <span name="Y">Y &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.fullOuterShuffleRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.fullOuterShuffleRegionJoin.Y">Y</span>]</span>, <span name="Z">Z &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.fullOuterShuffleRegionJoin.X">X</span>]), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.fullOuterShuffleRegionJoin.Z">Z</span>]</span>]</span><span class="params">(<span name="genomicRdd">genomicRdd: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.fullOuterShuffleRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.fullOuterShuffleRegionJoin.Y">Y</span>]</span>, <span name="optPartitions">optPartitions: <span class="extype" name="scala.Option">Option</span>[<span class="extype" name="scala.Int">Int</span>] = <span class="symbol">None</span></span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tTag">tTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="xTag">xTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.fullOuterShuffleRegionJoin.X">X</span>]</span>)</span><span class="result">: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.fullOuterShuffleRegionJoin.X">X</span>]), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.fullOuterShuffleRegionJoin.Z">Z</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">Performs a sort-merge full outer join between this RDD and another RDD.</p><div class="fullcomment"><div class="comment cmt"><p>Performs a sort-merge full outer join between this RDD and another RDD.</p><p>In a sort-merge join, both RDDs are co-partitioned and sorted. The
partitions are then zipped, and we do a merge join on each partition.
The key equality function used for this join is the reference region
overlap function. Since this is a full outer join, if a value from either
RDD does not overlap any values in the other RDD, it will be paired with
a <code>None</code> in the product of the join.
</p></div><dl class="paramcmts block"><dt class="param">genomicRdd</dt><dd class="cmt"><p>The right RDD in the join.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new genomic RDD containing all pairs of keys that
  overlapped in the genomic coordinate space, and values that did not
  overlap will be paired with a <code>None</code>.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="scala.AnyRef#getClass" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="getClass():Class[_]"></a>
      <a id="getClass():Class[_]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">getClass</span><span class="params">()</span><span class="result">: <span class="extype" name="java.lang.Class">Class</span>[_]</span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef → Any</dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#getReferenceRegions" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="getReferenceRegions(elem:org.bdgenomics.formats.avro.Genotype):Seq[org.bdgenomics.adam.models.ReferenceRegion]"></a>
      <a id="getReferenceRegions(Genotype):Seq[ReferenceRegion]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">getReferenceRegions</span><span class="params">(<span name="elem">elem: <span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span></span>)</span><span class="result">: <span class="extype" name="scala.Seq">Seq</span>[<a href="../../models/ReferenceRegion.html" class="extype" name="org.bdgenomics.adam.models.ReferenceRegion">ReferenceRegion</a>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt"></p><div class="fullcomment"><div class="comment cmt"></div><dl class="paramcmts block"><dt class="param">elem</dt><dd class="cmt"><p>The genotype to get a reference region for.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns the singular region this genotype covers.
</p></dd></dl><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd><a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a> → <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#headerLines" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="headerLines:Seq[htsjdk.variant.vcf.VCFHeaderLine]"></a>
      <a id="headerLines:Seq[VCFHeaderLine]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">val</span>
      </span>
      <span class="symbol">
        <span class="name">headerLines</span><span class="result">: <span class="extype" name="scala.Seq">Seq</span>[<span class="extype" name="htsjdk.variant.vcf.VCFHeaderLine">VCFHeaderLine</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">The VCF header lines that cover all INFO/FORMAT fields
  needed to represent this RDD of Genotypes.</p><div class="fullcomment"><div class="comment cmt"><p>The VCF header lines that cover all INFO/FORMAT fields
  needed to represent this RDD of Genotypes.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a> → <a href="../MultisampleAvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.MultisampleAvroGenomicRDD">MultisampleAvroGenomicRDD</a></dd></dl></div>
    </li><li name="scala.Any#isInstanceOf" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="isInstanceOf[T0]:Boolean"></a>
      <a id="isInstanceOf[T0]:Boolean"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">isInstanceOf</span><span class="tparams">[<span name="T0">T0</span>]</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>Any</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#isTraceEnabled" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="isTraceEnabled():Boolean"></a>
      <a id="isTraceEnabled():Boolean"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">isTraceEnabled</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#jrdd" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="jrdd:org.apache.spark.api.java.JavaRDD[T]"></a>
      <a id="jrdd:JavaRDD[Genotype]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">lazy val</span>
      </span>
      <span class="symbol">
        <span class="name">jrdd</span><span class="result">: <span class="extype" name="org.apache.spark.api.java.JavaRDD">JavaRDD</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">The underlying RDD of genomic data, as a JavaRDD.</p><div class="fullcomment"><div class="comment cmt"><p>The underlying RDD of genomic data, as a JavaRDD.
</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#leftOuterShuffleRegionJoin" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="leftOuterShuffleRegionJoin[X,Y&lt;:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],Z&lt;:org.bdgenomics.adam.rdd.GenomicRDD[(T,Option[X]),Z]](genomicRdd:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],optPartitions:Option[Int])(implicittTag:scala.reflect.ClassTag[T],implicitxTag:scala.reflect.ClassTag[X]):org.bdgenomics.adam.rdd.GenomicRDD[(T,Option[X]),Z]"></a>
      <a id="leftOuterShuffleRegionJoin[X,Y&lt;:GenomicRDD[X,Y],Z&lt;:GenomicRDD[(Genotype,Option[X]),Z]](GenomicRDD[X,Y],Option[Int])(ClassTag[Genotype],ClassTag[X]):GenomicRDD[(Genotype,Option[X]),Z]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">leftOuterShuffleRegionJoin</span><span class="tparams">[<span name="X">X</span>, <span name="Y">Y &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.leftOuterShuffleRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.leftOuterShuffleRegionJoin.Y">Y</span>]</span>, <span name="Z">Z &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.leftOuterShuffleRegionJoin.X">X</span>]), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.leftOuterShuffleRegionJoin.Z">Z</span>]</span>]</span><span class="params">(<span name="genomicRdd">genomicRdd: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.leftOuterShuffleRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.leftOuterShuffleRegionJoin.Y">Y</span>]</span>, <span name="optPartitions">optPartitions: <span class="extype" name="scala.Option">Option</span>[<span class="extype" name="scala.Int">Int</span>] = <span class="symbol">None</span></span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tTag">tTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="xTag">xTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.leftOuterShuffleRegionJoin.X">X</span>]</span>)</span><span class="result">: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.leftOuterShuffleRegionJoin.X">X</span>]), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.leftOuterShuffleRegionJoin.Z">Z</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">Performs a sort-merge left outer join between this RDD and another RDD.</p><div class="fullcomment"><div class="comment cmt"><p>Performs a sort-merge left outer join between this RDD and another RDD.</p><p>In a sort-merge join, both RDDs are co-partitioned and sorted. The
partitions are then zipped, and we do a merge join on each partition.
The key equality function used for this join is the reference region
overlap function. Since this is a left outer join, all values in the
right RDD that do not overlap a value from the left RDD are dropped.
If a value from the left RDD does not overlap any values in the right
RDD, it will be paired with a <code>None</code> in the product of the join.
</p></div><dl class="paramcmts block"><dt class="param">genomicRdd</dt><dd class="cmt"><p>The right RDD in the join.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new genomic RDD containing all pairs of keys that
  overlapped in the genomic coordinate space, and all keys from the
  left RDD that did not overlap a key in the right RDD.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#log" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="log:org.slf4j.Logger"></a>
      <a id="log:Logger"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">log</span><span class="result">: <span class="extype" name="org.slf4j.Logger">Logger</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#logDebug" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="logDebug(msg:=&gt;String,throwable:Throwable):Unit"></a>
      <a id="logDebug(⇒String,Throwable):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">logDebug</span><span class="params">(<span name="msg">msg: ⇒ <span class="extype" name="scala.Predef.String">String</span></span>, <span name="throwable">throwable: <span class="extype" name="scala.Throwable">Throwable</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#logDebug" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="logDebug(msg:=&gt;String):Unit"></a>
      <a id="logDebug(⇒String):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">logDebug</span><span class="params">(<span name="msg">msg: ⇒ <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#logError" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="logError(msg:=&gt;String,throwable:Throwable):Unit"></a>
      <a id="logError(⇒String,Throwable):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">logError</span><span class="params">(<span name="msg">msg: ⇒ <span class="extype" name="scala.Predef.String">String</span></span>, <span name="throwable">throwable: <span class="extype" name="scala.Throwable">Throwable</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#logError" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="logError(msg:=&gt;String):Unit"></a>
      <a id="logError(⇒String):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">logError</span><span class="params">(<span name="msg">msg: ⇒ <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#logInfo" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="logInfo(msg:=&gt;String,throwable:Throwable):Unit"></a>
      <a id="logInfo(⇒String,Throwable):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">logInfo</span><span class="params">(<span name="msg">msg: ⇒ <span class="extype" name="scala.Predef.String">String</span></span>, <span name="throwable">throwable: <span class="extype" name="scala.Throwable">Throwable</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#logInfo" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="logInfo(msg:=&gt;String):Unit"></a>
      <a id="logInfo(⇒String):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">logInfo</span><span class="params">(<span name="msg">msg: ⇒ <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#logName" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="logName:String"></a>
      <a id="logName:String"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">logName</span><span class="result">: <span class="extype" name="java.lang.String">String</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#logTrace" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="logTrace(msg:=&gt;String,throwable:Throwable):Unit"></a>
      <a id="logTrace(⇒String,Throwable):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">logTrace</span><span class="params">(<span name="msg">msg: ⇒ <span class="extype" name="scala.Predef.String">String</span></span>, <span name="throwable">throwable: <span class="extype" name="scala.Throwable">Throwable</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#logTrace" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="logTrace(msg:=&gt;String):Unit"></a>
      <a id="logTrace(⇒String):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">logTrace</span><span class="params">(<span name="msg">msg: ⇒ <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#logWarning" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="logWarning(msg:=&gt;String,throwable:Throwable):Unit"></a>
      <a id="logWarning(⇒String,Throwable):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">logWarning</span><span class="params">(<span name="msg">msg: ⇒ <span class="extype" name="scala.Predef.String">String</span></span>, <span name="throwable">throwable: <span class="extype" name="scala.Throwable">Throwable</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="org.bdgenomics.utils.misc.Logging#logWarning" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="logWarning(msg:=&gt;String):Unit"></a>
      <a id="logWarning(⇒String):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">logWarning</span><span class="params">(<span name="msg">msg: ⇒ <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>Logging</dd></dl></div>
    </li><li name="scala.AnyRef#ne" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="ne(x$1:AnyRef):Boolean"></a>
      <a id="ne(AnyRef):Boolean"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">ne</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.AnyRef">AnyRef</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="scala.AnyRef#notify" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="notify():Unit"></a>
      <a id="notify():Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">notify</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="scala.AnyRef#notifyAll" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="notifyAll():Unit"></a>
      <a id="notifyAll():Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">notifyAll</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#pipe" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="pipe[X,Y&lt;:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],V&lt;:org.bdgenomics.adam.rdd.InFormatter[T,U,V]](cmd:String,files:Seq[String],environment:Map[String,String],flankSize:Int)(implicittFormatterCompanion:org.bdgenomics.adam.rdd.InFormatterCompanion[T,U,V],implicitxFormatter:org.bdgenomics.adam.rdd.OutFormatter[X],implicitconvFn:(U,org.apache.spark.rdd.RDD[X])=&gt;Y,implicittManifest:scala.reflect.ClassTag[T],implicitxManifest:scala.reflect.ClassTag[X]):Y"></a>
      <a id="pipe[X,Y&lt;:GenomicRDD[X,Y],V&lt;:InFormatter[Genotype,GenotypeRDD,V]](String,Seq[String],Map[String,String],Int)(InFormatterCompanion[Genotype,GenotypeRDD,V],OutFormatter[X],(GenotypeRDD,RDD[X])⇒Y,ClassTag[Genotype],ClassTag[X]):Y"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">pipe</span><span class="tparams">[<span name="X">X</span>, <span name="Y">Y &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.pipe.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.pipe.Y">Y</span>]</span>, <span name="V">V &lt;: <a href="../InFormatter.html" class="extype" name="org.bdgenomics.adam.rdd.InFormatter">InFormatter</a>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.pipe.V">V</span>]</span>]</span><span class="params">(<span name="cmd">cmd: <span class="extype" name="scala.Predef.String">String</span></span>, <span name="files">files: <span class="extype" name="scala.Seq">Seq</span>[<span class="extype" name="scala.Predef.String">String</span>] = <span class="symbol">Seq.empty</span></span>, <span name="environment">environment: <span class="extype" name="scala.Predef.Map">Map</span>[<span class="extype" name="scala.Predef.String">String</span>, <span class="extype" name="scala.Predef.String">String</span>] = <span class="symbol">Map.empty</span></span>, <span name="flankSize">flankSize: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">0</span></span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tFormatterCompanion">tFormatterCompanion: <a href="../InFormatterCompanion.html" class="extype" name="org.bdgenomics.adam.rdd.InFormatterCompanion">InFormatterCompanion</a>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.pipe.V">V</span>]</span>, <span name="xFormatter">xFormatter: <a href="../OutFormatter.html" class="extype" name="org.bdgenomics.adam.rdd.OutFormatter">OutFormatter</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.pipe.X">X</span>]</span>, <span name="convFn">convFn: (<a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>, <span class="extype" name="org.apache.spark.rdd.RDD">RDD</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.pipe.X">X</span>]) ⇒ <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.pipe.Y">Y</span></span>, <span name="tManifest">tManifest: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="xManifest">xManifest: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.pipe.X">X</span>]</span>)</span><span class="result">: <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.pipe.Y">Y</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Pipes genomic data to a subprocess that runs in parallel using Spark.</p><div class="fullcomment"><div class="comment cmt"><p>Pipes genomic data to a subprocess that runs in parallel using Spark.</p><p>Files are substituted in to the command with a $x syntax. E.g., to invoke
a command that uses the first file from the files Seq, use $0.</p><p>Pipes require the presence of an InFormatterCompanion and an OutFormatter
as implicit values. The InFormatterCompanion should be a singleton whose
apply method builds an InFormatter given a specific type of GenomicRDD.
The implicit InFormatterCompanion yields an InFormatter which is used to
format the input to the pipe, and the implicit OutFormatter is used to
parse the output from the pipe.
</p></div><dl class="paramcmts block"><dt class="tparam">X</dt><dd class="cmt"><p>The type of the record created by the piped command.</p></dd><dt class="tparam">Y</dt><dd class="cmt"><p>A GenomicRDD containing X's.</p></dd><dt class="tparam">V</dt><dd class="cmt"><p>The InFormatter to use for formatting the data being piped to the
  command.
</p></dd><dt class="param">cmd</dt><dd class="cmt"><p>Command to run.</p></dd><dt class="param">files</dt><dd class="cmt"><p>Files to make locally available to the commands being run.
  Default is empty.</p></dd><dt class="param">environment</dt><dd class="cmt"><p>A map containing environment variable/value pairs to set
  in the environment for the newly created process. Default is empty.</p></dd><dt class="param">flankSize</dt><dd class="cmt"><p>Number of bases to flank each command invocation by.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new GenomicRDD of type Y.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#rdd" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="rdd:org.apache.spark.rdd.RDD[org.bdgenomics.formats.avro.Genotype]"></a>
      <a id="rdd:RDD[Genotype]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">val</span>
      </span>
      <span class="symbol">
        <span class="name">rdd</span><span class="result">: <span class="extype" name="org.apache.spark.rdd.RDD">RDD</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">Called genotypes.</p><div class="fullcomment"><div class="comment cmt"><p>Called genotypes.</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a> → <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a> → ADAMRDDFunctions</dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#replaceRdd" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="replaceRdd(newRdd:org.apache.spark.rdd.RDD[org.bdgenomics.formats.avro.Genotype]):org.bdgenomics.adam.rdd.variant.GenotypeRDD"></a>
      <a id="replaceRdd(RDD[Genotype]):GenotypeRDD"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">replaceRdd</span><span class="params">(<span name="newRdd">newRdd: <span class="extype" name="org.apache.spark.rdd.RDD">RDD</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>)</span><span class="result">: <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a></span>
      </span>
      </h4>
      <p class="shortcomment cmt"></p><div class="fullcomment"><div class="comment cmt"></div><dl class="paramcmts block"><dt class="param">newRdd</dt><dd class="cmt"><p>An RDD to replace the underlying RDD with.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new GenotypeRDD with the underlying RDD replaced.
</p></dd></dl><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd><a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a> → <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#rightOuterBroadcastRegionJoin" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="rightOuterBroadcastRegionJoin[X,Y&lt;:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],Z&lt;:org.bdgenomics.adam.rdd.GenomicRDD[(Option[T],X),Z]](genomicRdd:org.bdgenomics.adam.rdd.GenomicRDD[X,Y])(implicittTag:scala.reflect.ClassTag[T],implicitxTag:scala.reflect.ClassTag[X]):org.bdgenomics.adam.rdd.GenomicRDD[(Option[T],X),Z]"></a>
      <a id="rightOuterBroadcastRegionJoin[X,Y&lt;:GenomicRDD[X,Y],Z&lt;:GenomicRDD[(Option[Genotype],X),Z]](GenomicRDD[X,Y])(ClassTag[Genotype],ClassTag[X]):GenomicRDD[(Option[Genotype],X),Z]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">rightOuterBroadcastRegionJoin</span><span class="tparams">[<span name="X">X</span>, <span name="Y">Y &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoin.Y">Y</span>]</span>, <span name="Z">Z &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoin.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoin.Z">Z</span>]</span>]</span><span class="params">(<span name="genomicRdd">genomicRdd: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoin.Y">Y</span>]</span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tTag">tTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="xTag">xTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoin.X">X</span>]</span>)</span><span class="result">: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoin.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoin.Z">Z</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">Performs a broadcast right outer join between this RDD and another RDD.</p><div class="fullcomment"><div class="comment cmt"><p>Performs a broadcast right outer join between this RDD and another RDD.</p><p>In a broadcast join, the left RDD (this RDD) is collected to the driver,
and broadcast to all the nodes in the cluster. The key equality function
used for this join is the reference region overlap function. Since this
is a right outer join, all values in the left RDD that do not overlap a
value from the right RDD are dropped. If a value from the right RDD does
not overlap any values in the left RDD, it will be paired with a <code>None</code>
in the product of the join.
</p></div><dl class="paramcmts block"><dt class="param">genomicRdd</dt><dd class="cmt"><p>The right RDD in the join.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new genomic RDD containing all pairs of keys that
  overlapped in the genomic coordinate space, and all keys from the
  right RDD that did not overlap a key in the left RDD.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#rightOuterBroadcastRegionJoinAndGroupByRight" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="rightOuterBroadcastRegionJoinAndGroupByRight[X,Y&lt;:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],Z&lt;:org.bdgenomics.adam.rdd.GenomicRDD[(Iterable[T],X),Z]](genomicRdd:org.bdgenomics.adam.rdd.GenomicRDD[X,Y])(implicittTag:scala.reflect.ClassTag[T],implicitxTag:scala.reflect.ClassTag[X]):org.bdgenomics.adam.rdd.GenomicRDD[(Iterable[T],X),Z]"></a>
      <a id="rightOuterBroadcastRegionJoinAndGroupByRight[X,Y&lt;:GenomicRDD[X,Y],Z&lt;:GenomicRDD[(Iterable[Genotype],X),Z]](GenomicRDD[X,Y])(ClassTag[Genotype],ClassTag[X]):GenomicRDD[(Iterable[Genotype],X),Z]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">rightOuterBroadcastRegionJoinAndGroupByRight</span><span class="tparams">[<span name="X">X</span>, <span name="Y">Y &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoinAndGroupByRight.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoinAndGroupByRight.Y">Y</span>]</span>, <span name="Z">Z &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Iterable">Iterable</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoinAndGroupByRight.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoinAndGroupByRight.Z">Z</span>]</span>]</span><span class="params">(<span name="genomicRdd">genomicRdd: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoinAndGroupByRight.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoinAndGroupByRight.Y">Y</span>]</span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tTag">tTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="xTag">xTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoinAndGroupByRight.X">X</span>]</span>)</span><span class="result">: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Iterable">Iterable</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoinAndGroupByRight.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterBroadcastRegionJoinAndGroupByRight.Z">Z</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">Performs a broadcast right outer join between this RDD and another RDD.</p><div class="fullcomment"><div class="comment cmt"><p>Performs a broadcast right outer join between this RDD and another RDD.</p><p>In a broadcast join, the left RDD (this RDD) is collected to the driver,
and broadcast to all the nodes in the cluster. The key equality function
used for this join is the reference region overlap function. Since this
is a right outer join, all values in the left RDD that do not overlap a
value from the right RDD are dropped. If a value from the right RDD does
not overlap any values in the left RDD, it will be paired with a <code>None</code>
in the product of the join.
</p></div><dl class="paramcmts block"><dt class="param">genomicRdd</dt><dd class="cmt"><p>The right RDD in the join.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new genomic RDD containing all pairs of keys that
  overlapped in the genomic coordinate space, and all keys from the
  right RDD that did not overlap a key in the left RDD.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#rightOuterShuffleRegionJoin" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="rightOuterShuffleRegionJoin[X,Y&lt;:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],Z&lt;:org.bdgenomics.adam.rdd.GenomicRDD[(Option[T],X),Z]](genomicRdd:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],optPartitions:Option[Int])(implicittTag:scala.reflect.ClassTag[T],implicitxTag:scala.reflect.ClassTag[X]):org.bdgenomics.adam.rdd.GenomicRDD[(Option[T],X),Z]"></a>
      <a id="rightOuterShuffleRegionJoin[X,Y&lt;:GenomicRDD[X,Y],Z&lt;:GenomicRDD[(Option[Genotype],X),Z]](GenomicRDD[X,Y],Option[Int])(ClassTag[Genotype],ClassTag[X]):GenomicRDD[(Option[Genotype],X),Z]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">rightOuterShuffleRegionJoin</span><span class="tparams">[<span name="X">X</span>, <span name="Y">Y &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoin.Y">Y</span>]</span>, <span name="Z">Z &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoin.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoin.Z">Z</span>]</span>]</span><span class="params">(<span name="genomicRdd">genomicRdd: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoin.Y">Y</span>]</span>, <span name="optPartitions">optPartitions: <span class="extype" name="scala.Option">Option</span>[<span class="extype" name="scala.Int">Int</span>] = <span class="symbol">None</span></span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tTag">tTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="xTag">xTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoin.X">X</span>]</span>)</span><span class="result">: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoin.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoin.Z">Z</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">Performs a sort-merge right outer join between this RDD and another RDD.</p><div class="fullcomment"><div class="comment cmt"><p>Performs a sort-merge right outer join between this RDD and another RDD.</p><p>In a sort-merge join, both RDDs are co-partitioned and sorted. The
partitions are then zipped, and we do a merge join on each partition.
The key equality function used for this join is the reference region
overlap function. Since this is a right outer join, all values in the
left RDD that do not overlap a value from the right RDD are dropped.
If a value from the right RDD does not overlap any values in the left
RDD, it will be paired with a <code>None</code> in the product of the join.
</p></div><dl class="paramcmts block"><dt class="param">genomicRdd</dt><dd class="cmt"><p>The right RDD in the join.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new genomic RDD containing all pairs of keys that
  overlapped in the genomic coordinate space, and all keys from the
  right RDD that did not overlap a key in the left RDD.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#rightOuterShuffleRegionJoinAndGroupByLeft" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="rightOuterShuffleRegionJoinAndGroupByLeft[X,Y&lt;:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],Z&lt;:org.bdgenomics.adam.rdd.GenomicRDD[(Option[T],Iterable[X]),Z]](genomicRdd:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],optPartitions:Option[Int])(implicittTag:scala.reflect.ClassTag[T],implicitxTag:scala.reflect.ClassTag[X]):org.bdgenomics.adam.rdd.GenomicRDD[(Option[T],Iterable[X]),Z]"></a>
      <a id="rightOuterShuffleRegionJoinAndGroupByLeft[X,Y&lt;:GenomicRDD[X,Y],Z&lt;:GenomicRDD[(Option[Genotype],Iterable[X]),Z]](GenomicRDD[X,Y],Option[Int])(ClassTag[Genotype],ClassTag[X]):GenomicRDD[(Option[Genotype],Iterable[X]),Z]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">rightOuterShuffleRegionJoinAndGroupByLeft</span><span class="tparams">[<span name="X">X</span>, <span name="Y">Y &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoinAndGroupByLeft.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoinAndGroupByLeft.Y">Y</span>]</span>, <span name="Z">Z &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="scala.Iterable">Iterable</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoinAndGroupByLeft.X">X</span>]), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoinAndGroupByLeft.Z">Z</span>]</span>]</span><span class="params">(<span name="genomicRdd">genomicRdd: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoinAndGroupByLeft.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoinAndGroupByLeft.Y">Y</span>]</span>, <span name="optPartitions">optPartitions: <span class="extype" name="scala.Option">Option</span>[<span class="extype" name="scala.Int">Int</span>] = <span class="symbol">None</span></span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tTag">tTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="xTag">xTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoinAndGroupByLeft.X">X</span>]</span>)</span><span class="result">: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>], <span class="extype" name="scala.Iterable">Iterable</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoinAndGroupByLeft.X">X</span>]), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.rightOuterShuffleRegionJoinAndGroupByLeft.Z">Z</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">Performs a sort-merge right outer join between this RDD and another RDD,
followed by a groupBy on the left value, if not null.</p><div class="fullcomment"><div class="comment cmt"><p>Performs a sort-merge right outer join between this RDD and another RDD,
followed by a groupBy on the left value, if not null.</p><p>In a sort-merge join, both RDDs are co-partitioned and sorted. The
partitions are then zipped, and we do a merge join on each partition.
The key equality function used for this join is the reference region
overlap function. In the same operation, we group all values by the left
item in the RDD. Since this is a right outer join, all values from the
right RDD who did not overlap a value from the left RDD are placed into
a length-1 Iterable with a <code>None</code> key.
</p></div><dl class="paramcmts block"><dt class="param">genomicRdd</dt><dd class="cmt"><p>The right RDD in the join.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new genomic RDD containing all pairs of keys that
  overlapped in the genomic coordinate space, grouped together by
  the value they overlapped in the left RDD, and all values from the
  right RDD that did not overlap an item in the left RDD.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#samples" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="samples:Seq[org.bdgenomics.formats.avro.Sample]"></a>
      <a id="samples:Seq[Sample]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">val</span>
      </span>
      <span class="symbol">
        <span class="name">samples</span><span class="result">: <span class="extype" name="scala.Seq">Seq</span>[<span class="extype" name="org.bdgenomics.formats.avro.Sample">Sample</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">The samples called.</p><div class="fullcomment"><div class="comment cmt"><p>The samples called.</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a> → <a href="../MultisampleGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.MultisampleGenomicRDD">MultisampleGenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#save" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="save(args:org.bdgenomics.utils.cli.SaveArgs):Boolean"></a>
      <a id="save(SaveArgs):Boolean"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">save</span><span class="params">(<span name="args">args: <span class="extype" name="org.bdgenomics.utils.cli.SaveArgs">SaveArgs</span></span>)</span><span class="result">: <span class="extype" name="scala.Boolean">Boolean</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Automatically detects the extension and saves to either VCF or Parquet.</p><div class="fullcomment"><div class="comment cmt"><p>Automatically detects the extension and saves to either VCF or Parquet.
</p></div><dl class="paramcmts block"><dt class="param">args</dt><dd class="cmt"><p>Arguments configuring how to save the output.
</p></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#save" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="save(filePath:String):Unit"></a>
      <a id="save(String):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">save</span><span class="params">(<span name="filePath">filePath: <span class="extype" name="java.lang.String">String</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Java-friendly method for saving.</p><div class="fullcomment"><div class="comment cmt"><p>Java-friendly method for saving.
</p></div><dl class="paramcmts block"><dt class="param">filePath</dt><dd class="cmt"><p>Path to save file to. If ends in &quot;.vcf&quot;, saves as VCF, else
  saves as Parquet.
</p></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.AvroGenomicRDD#saveAsParquet" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="saveAsParquet(filePath:String):Unit"></a>
      <a id="saveAsParquet(String):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">saveAsParquet</span><span class="params">(<span name="filePath">filePath: <span class="extype" name="java.lang.String">String</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Saves this RDD to disk as a Parquet file.</p><div class="fullcomment"><div class="comment cmt"><p>Saves this RDD to disk as a Parquet file.
</p></div><dl class="paramcmts block"><dt class="param">filePath</dt><dd class="cmt"><p>Path to save the file at.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../AvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.AvroGenomicRDD">AvroGenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.AvroGenomicRDD#saveAsParquet" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="saveAsParquet(filePath:String,blockSize:Integer,pageSize:Integer,compressCodec:org.apache.parquet.hadoop.metadata.CompressionCodecName,disableDictionaryEncoding:Boolean):Unit"></a>
      <a id="saveAsParquet(String,Integer,Integer,CompressionCodecName,Boolean):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">saveAsParquet</span><span class="params">(<span name="filePath">filePath: <span class="extype" name="java.lang.String">String</span></span>, <span name="blockSize">blockSize: <span class="extype" name="java.lang.Integer">Integer</span></span>, <span name="pageSize">pageSize: <span class="extype" name="java.lang.Integer">Integer</span></span>, <span name="compressCodec">compressCodec: <span class="extype" name="org.apache.parquet.hadoop.metadata.CompressionCodecName">CompressionCodecName</span></span>, <span name="disableDictionaryEncoding">disableDictionaryEncoding: <span class="extype" name="java.lang.Boolean">Boolean</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Saves this RDD to disk as a Parquet file.</p><div class="fullcomment"><div class="comment cmt"><p>Saves this RDD to disk as a Parquet file.
</p></div><dl class="paramcmts block"><dt class="param">filePath</dt><dd class="cmt"><p>Path to save the file at.</p></dd><dt class="param">blockSize</dt><dd class="cmt"><p>Size per block.</p></dd><dt class="param">pageSize</dt><dd class="cmt"><p>Size per page.</p></dd><dt class="param">compressCodec</dt><dd class="cmt"><p>Name of the compression codec to use.</p></dd><dt class="param">disableDictionaryEncoding</dt><dd class="cmt"><p>Whether or not to disable bit-packing.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../AvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.AvroGenomicRDD">AvroGenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.AvroGenomicRDD#saveAsParquet" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="saveAsParquet(filePath:String,blockSize:Int,pageSize:Int,compressCodec:org.apache.parquet.hadoop.metadata.CompressionCodecName,disableDictionaryEncoding:Boolean):Unit"></a>
      <a id="saveAsParquet(String,Int,Int,CompressionCodecName,Boolean):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">saveAsParquet</span><span class="params">(<span name="filePath">filePath: <span class="extype" name="scala.Predef.String">String</span></span>, <span name="blockSize">blockSize: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">128 * 1024 * 1024</span></span>, <span name="pageSize">pageSize: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">1 * 1024 * 1024</span></span>, <span name="compressCodec">compressCodec: <span class="extype" name="org.apache.parquet.hadoop.metadata.CompressionCodecName">CompressionCodecName</span> = <span class="symbol"><span class="name"><a href="../../../../package.html">CompressionCodecName.GZIP</a></span></span></span>, <span name="disableDictionaryEncoding">disableDictionaryEncoding: <span class="extype" name="scala.Boolean">Boolean</span> = <span class="symbol">false</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Saves this RDD to disk as a Parquet file.</p><div class="fullcomment"><div class="comment cmt"><p>Saves this RDD to disk as a Parquet file.
</p></div><dl class="paramcmts block"><dt class="param">filePath</dt><dd class="cmt"><p>Path to save the file at.</p></dd><dt class="param">blockSize</dt><dd class="cmt"><p>Size per block.</p></dd><dt class="param">pageSize</dt><dd class="cmt"><p>Size per page.</p></dd><dt class="param">compressCodec</dt><dd class="cmt"><p>Name of the compression codec to use.</p></dd><dt class="param">disableDictionaryEncoding</dt><dd class="cmt"><p>Whether or not to disable bit-packing.
  Default is false.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../AvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.AvroGenomicRDD">AvroGenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.AvroGenomicRDD#saveAsParquet" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="saveAsParquet(args:org.bdgenomics.utils.cli.SaveArgs):Unit"></a>
      <a id="saveAsParquet(SaveArgs):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">saveAsParquet</span><span class="params">(<span name="args">args: <span class="extype" name="org.bdgenomics.utils.cli.SaveArgs">SaveArgs</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Saves RDD as a directory of Parquet files.</p><div class="fullcomment"><div class="comment cmt"><p>Saves RDD as a directory of Parquet files.</p><p>The RDD is written as a directory of Parquet files, with
Parquet configuration described by the input param args.
The provided sequence dictionary is written at args.outputPath/_seqdict.avro
as Avro binary.
</p></div><dl class="paramcmts block"><dt class="param">args</dt><dd class="cmt"><p>Save configuration arguments.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../AvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.AvroGenomicRDD">AvroGenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#saveAsVcf" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="saveAsVcf(filePath:String,asSingleFile:Boolean,stringency:htsjdk.samtools.ValidationStringency):Unit"></a>
      <a id="saveAsVcf(String,Boolean,ValidationStringency):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">saveAsVcf</span><span class="params">(<span name="filePath">filePath: <span class="extype" name="scala.Predef.String">String</span></span>, <span name="asSingleFile">asSingleFile: <span class="extype" name="scala.Boolean">Boolean</span></span>, <span name="stringency">stringency: <span class="extype" name="htsjdk.samtools.ValidationStringency">ValidationStringency</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Explicitly saves to VCF.</p><div class="fullcomment"><div class="comment cmt"><p>Explicitly saves to VCF.
</p></div><dl class="paramcmts block"><dt class="param">filePath</dt><dd class="cmt"><p>The filepath to save to.</p></dd><dt class="param">asSingleFile</dt><dd class="cmt"><p>If true, saves the output as a single file by merging
  the sharded output after completing the write to HDFS. If false, the
  output of this call will be written as shards, where each shard has a
  valid VCF header.</p></dd><dt class="param">stringency</dt><dd class="cmt"><p>The validation stringency to use when writing the VCF.
</p></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#saveAsVcf" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="saveAsVcf(args:org.bdgenomics.utils.cli.SaveArgs,sortOnSave:Boolean):Unit"></a>
      <a id="saveAsVcf(SaveArgs,Boolean):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">saveAsVcf</span><span class="params">(<span name="args">args: <span class="extype" name="org.bdgenomics.utils.cli.SaveArgs">SaveArgs</span></span>, <span name="sortOnSave">sortOnSave: <span class="extype" name="scala.Boolean">Boolean</span> = <span class="symbol">false</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Explicitly saves to VCF.</p><div class="fullcomment"><div class="comment cmt"><p>Explicitly saves to VCF.
</p></div><dl class="paramcmts block"><dt class="param">args</dt><dd class="cmt"><p>Arguments configuring how/where to save the output.</p></dd><dt class="param">sortOnSave</dt><dd class="cmt"><p>Whether to sort when saving or not.
</p></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.ADAMRDDFunctions#saveAvro" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="saveAvro[U&lt;:org.apache.avro.specific.SpecificRecordBase](filename:String,sc:org.apache.spark.SparkContext,schema:org.apache.avro.Schema,avro:Seq[U])(implicittUag:scala.reflect.ClassTag[U]):Unit"></a>
      <a id="saveAvro[U&lt;:SpecificRecordBase](String,SparkContext,Schema,Seq[U])(ClassTag[U]):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">saveAvro</span><span class="tparams">[<span name="U">U &lt;: <span class="extype" name="org.apache.avro.specific.SpecificRecordBase">SpecificRecordBase</span></span>]</span><span class="params">(<span name="filename">filename: <span class="extype" name="scala.Predef.String">String</span></span>, <span name="sc">sc: <span class="extype" name="org.apache.spark.SparkContext">SparkContext</span></span>, <span name="schema">schema: <span class="extype" name="org.apache.avro.Schema">Schema</span></span>, <span name="avro">avro: <span class="extype" name="scala.Seq">Seq</span>[<span class="extype" name="org.bdgenomics.adam.rdd.ADAMRDDFunctions.saveAvro.U">U</span>]</span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tUag">tUag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.ADAMRDDFunctions.saveAvro.U">U</span>]</span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Saves Avro data to a Hadoop file system.</p><div class="fullcomment"><div class="comment cmt"><p>Saves Avro data to a Hadoop file system.</p><p>This method uses a SparkContext to identify our underlying file system,
which we then save to.</p><p>Frustratingly enough, although all records generated by the Avro IDL
compiler have a static SCHEMA$ field, this field does not belong to
the SpecificRecordBase abstract class, or the SpecificRecord interface.
As such, we must force the user to pass in the schema.
</p></div><dl class="paramcmts block"><dt class="tparam">U</dt><dd class="cmt"><p>The type of the specific record we are saving.</p></dd><dt class="param">filename</dt><dd class="cmt"><p>Path to save records to.</p></dd><dt class="param">sc</dt><dd class="cmt"><p>SparkContext used for identifying underlying file system.</p></dd><dt class="param">schema</dt><dd class="cmt"><p>Schema of records we are saving.</p></dd><dt class="param">avro</dt><dd class="cmt"><p>Seq of records we are saving.
</p></dd></dl><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>ADAMRDDFunctions</dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.MultisampleAvroGenomicRDD#saveMetadata" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="saveMetadata(filePath:String):Unit"></a>
      <a id="saveMetadata(String):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">saveMetadata</span><span class="params">(<span name="filePath">filePath: <span class="extype" name="scala.Predef.String">String</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Called in saveAsParquet after saving RDD to Parquet to save metadata.</p><div class="fullcomment"><div class="comment cmt"><p>Called in saveAsParquet after saving RDD to Parquet to save metadata.</p><p>Writes any necessary metadata to disk. If not overridden, writes the
sequence dictionary to disk as Avro.
</p></div><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd><a href="../MultisampleAvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.MultisampleAvroGenomicRDD">MultisampleAvroGenomicRDD</a> → <a href="../AvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.AvroGenomicRDD">AvroGenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.ADAMRDDFunctions#saveRddAsParquet" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="saveRddAsParquet(filePath:String,blockSize:Int,pageSize:Int,compressCodec:org.apache.parquet.hadoop.metadata.CompressionCodecName,disableDictionaryEncoding:Boolean,schema:Option[org.apache.avro.Schema]):Unit"></a>
      <a id="saveRddAsParquet(String,Int,Int,CompressionCodecName,Boolean,Option[Schema]):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">saveRddAsParquet</span><span class="params">(<span name="filePath">filePath: <span class="extype" name="scala.Predef.String">String</span></span>, <span name="blockSize">blockSize: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">128 * 1024 * 1024</span></span>, <span name="pageSize">pageSize: <span class="extype" name="scala.Int">Int</span> = <span class="symbol">1 * 1024 * 1024</span></span>, <span name="compressCodec">compressCodec: <span class="extype" name="org.apache.parquet.hadoop.metadata.CompressionCodecName">CompressionCodecName</span> = <span class="symbol"><span class="name"><a href="../../../../package.html">CompressionCodecName.GZIP</a></span></span></span>, <span name="disableDictionaryEncoding">disableDictionaryEncoding: <span class="extype" name="scala.Boolean">Boolean</span> = <span class="symbol">false</span></span>, <span name="schema">schema: <span class="extype" name="scala.Option">Option</span>[<span class="extype" name="org.apache.avro.Schema">Schema</span>] = <span class="symbol">None</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Saves an RDD of Avro data to Parquet.</p><div class="fullcomment"><div class="comment cmt"><p>Saves an RDD of Avro data to Parquet.
</p></div><dl class="paramcmts block"><dt class="param">filePath</dt><dd class="cmt"><p>The path to save the file to.</p></dd><dt class="param">blockSize</dt><dd class="cmt"><p>The size in bytes of blocks to write.</p></dd><dt class="param">pageSize</dt><dd class="cmt"><p>The size in bytes of pages to write.</p></dd><dt class="param">compressCodec</dt><dd class="cmt"><p>The compression codec to apply to pages.</p></dd><dt class="param">disableDictionaryEncoding</dt><dd class="cmt"><p>If false, dictionary encoding is used. If
  true, delta encoding is used.</p></dd><dt class="param">schema</dt><dd class="cmt"><p>The schema to set.
</p></dd></dl><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>ADAMRDDFunctions</dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.ADAMRDDFunctions#saveRddAsParquet" visbl="prt" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="saveRddAsParquet(args:org.bdgenomics.utils.cli.SaveArgs):Unit"></a>
      <a id="saveRddAsParquet(SaveArgs):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">saveRddAsParquet</span><span class="params">(<span name="args">args: <span class="extype" name="org.bdgenomics.utils.cli.SaveArgs">SaveArgs</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Attributes</dt><dd>protected </dd><dt>Definition Classes</dt><dd>ADAMRDDFunctions</dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#sequences" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="sequences:org.bdgenomics.adam.models.SequenceDictionary"></a>
      <a id="sequences:SequenceDictionary"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">val</span>
      </span>
      <span class="symbol">
        <span class="name">sequences</span><span class="result">: <a href="../../models/SequenceDictionary.html" class="extype" name="org.bdgenomics.adam.models.SequenceDictionary">SequenceDictionary</a></span>
      </span>
      </h4>
      <p class="shortcomment cmt">A dictionary describing the reference genome.</p><div class="fullcomment"><div class="comment cmt"><p>A dictionary describing the reference genome.</p></div><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a> → <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#shuffleRegionJoin" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="shuffleRegionJoin[X,Y&lt;:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],Z&lt;:org.bdgenomics.adam.rdd.GenomicRDD[(T,X),Z]](genomicRdd:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],optPartitions:Option[Int])(implicittTag:scala.reflect.ClassTag[T],implicitxTag:scala.reflect.ClassTag[X]):org.bdgenomics.adam.rdd.GenomicRDD[(T,X),Z]"></a>
      <a id="shuffleRegionJoin[X,Y&lt;:GenomicRDD[X,Y],Z&lt;:GenomicRDD[(Genotype,X),Z]](GenomicRDD[X,Y],Option[Int])(ClassTag[Genotype],ClassTag[X]):GenomicRDD[(Genotype,X),Z]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">shuffleRegionJoin</span><span class="tparams">[<span name="X">X</span>, <span name="Y">Y &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoin.Y">Y</span>]</span>, <span name="Z">Z &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoin.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoin.Z">Z</span>]</span>]</span><span class="params">(<span name="genomicRdd">genomicRdd: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoin.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoin.Y">Y</span>]</span>, <span name="optPartitions">optPartitions: <span class="extype" name="scala.Option">Option</span>[<span class="extype" name="scala.Int">Int</span>] = <span class="symbol">None</span></span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tTag">tTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="xTag">xTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoin.X">X</span>]</span>)</span><span class="result">: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoin.X">X</span>), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoin.Z">Z</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">Performs a sort-merge inner join between this RDD and another RDD.</p><div class="fullcomment"><div class="comment cmt"><p>Performs a sort-merge inner join between this RDD and another RDD.</p><p>In a sort-merge join, both RDDs are co-partitioned and sorted. The
partitions are then zipped, and we do a merge join on each partition.
The key equality function used for this join is the reference region
overlap function. Since this is an inner join, all values who do not
overlap a value from the other RDD are dropped.
</p></div><dl class="paramcmts block"><dt class="param">genomicRdd</dt><dd class="cmt"><p>The right RDD in the join.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new genomic RDD containing all pairs of keys that
  overlapped in the genomic coordinate space.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#shuffleRegionJoinAndGroupByLeft" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="shuffleRegionJoinAndGroupByLeft[X,Y&lt;:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],Z&lt;:org.bdgenomics.adam.rdd.GenomicRDD[(T,Iterable[X]),Z]](genomicRdd:org.bdgenomics.adam.rdd.GenomicRDD[X,Y],optPartitions:Option[Int])(implicittTag:scala.reflect.ClassTag[T],implicitxTag:scala.reflect.ClassTag[X]):org.bdgenomics.adam.rdd.GenomicRDD[(T,Iterable[X]),Z]"></a>
      <a id="shuffleRegionJoinAndGroupByLeft[X,Y&lt;:GenomicRDD[X,Y],Z&lt;:GenomicRDD[(Genotype,Iterable[X]),Z]](GenomicRDD[X,Y],Option[Int])(ClassTag[Genotype],ClassTag[X]):GenomicRDD[(Genotype,Iterable[X]),Z]"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">shuffleRegionJoinAndGroupByLeft</span><span class="tparams">[<span name="X">X</span>, <span name="Y">Y &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoinAndGroupByLeft.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoinAndGroupByLeft.Y">Y</span>]</span>, <span name="Z">Z &lt;: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <span class="extype" name="scala.Iterable">Iterable</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoinAndGroupByLeft.X">X</span>]), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoinAndGroupByLeft.Z">Z</span>]</span>]</span><span class="params">(<span name="genomicRdd">genomicRdd: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoinAndGroupByLeft.X">X</span>, <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoinAndGroupByLeft.Y">Y</span>]</span>, <span name="optPartitions">optPartitions: <span class="extype" name="scala.Option">Option</span>[<span class="extype" name="scala.Int">Int</span>] = <span class="symbol">None</span></span>)</span><span class="params">(<span class="implicit">implicit </span><span name="tTag">tTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>, <span name="xTag">xTag: <span class="extype" name="scala.reflect.ClassTag">ClassTag</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoinAndGroupByLeft.X">X</span>]</span>)</span><span class="result">: <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[(<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <span class="extype" name="scala.Iterable">Iterable</span>[<span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoinAndGroupByLeft.X">X</span>]), <span class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD.shuffleRegionJoinAndGroupByLeft.Z">Z</span>]</span>
      </span>
      </h4>
      <p class="shortcomment cmt">Performs a sort-merge inner join between this RDD and another RDD,
followed by a groupBy on the left value.</p><div class="fullcomment"><div class="comment cmt"><p>Performs a sort-merge inner join between this RDD and another RDD,
followed by a groupBy on the left value.</p><p>In a sort-merge join, both RDDs are co-partitioned and sorted. The
partitions are then zipped, and we do a merge join on each partition.
The key equality function used for this join is the reference region
overlap function. Since this is an inner join, all values who do not
overlap a value from the other RDD are dropped. In the same operation,
we group all values by the left item in the RDD.
</p></div><dl class="paramcmts block"><dt class="param">genomicRdd</dt><dd class="cmt"><p>The right RDD in the join.</p></dd><dt>returns</dt><dd class="cmt"><p>Returns a new genomic RDD containing all pairs of keys that
  overlapped in the genomic coordinate space, grouped together by
  the value they overlapped in the left RDD..
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#sort" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="sort():U"></a>
      <a id="sort():GenotypeRDD"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">sort</span><span class="params">()</span><span class="result">: <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Sorts our genome aligned data by reference positions, with contigs ordered
by index.</p><div class="fullcomment"><div class="comment cmt"><p>Sorts our genome aligned data by reference positions, with contigs ordered
by index.
</p></div><dl class="paramcmts block"><dt>returns</dt><dd class="cmt"><p>Returns a new RDD containing sorted data.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd><dt>Note</dt><dd><span class="cmt"><p>Does not support data that is unaligned or where objects align to
  multiple positions.</p></span></dd><dt>See also</dt><dd><span class="cmt"><p>sortLexicographically
</p></span></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#sortLexicographically" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="sortLexicographically():U"></a>
      <a id="sortLexicographically():GenotypeRDD"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">sortLexicographically</span><span class="params">()</span><span class="result">: <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Sorts our genome aligned data by reference positions, with contigs ordered
lexicographically.</p><div class="fullcomment"><div class="comment cmt"><p>Sorts our genome aligned data by reference positions, with contigs ordered
lexicographically.
</p></div><dl class="paramcmts block"><dt>returns</dt><dd class="cmt"><p>Returns a new RDD containing sorted data.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd><dt>Note</dt><dd><span class="cmt"><p>Does not support data that is unaligned or where objects align to
  multiple positions.</p></span></dd><dt>See also</dt><dd><span class="cmt"><p>sort
</p></span></dd></dl></div>
    </li><li name="scala.AnyRef#synchronized" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="synchronized[T0](x$1:=&gt;T0):T0"></a>
      <a id="synchronized[T0](⇒T0):T0"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">synchronized</span><span class="tparams">[<span name="T0">T0</span>]</span><span class="params">(<span name="arg0">arg0: ⇒ <span class="extype" name="java.lang.AnyRef.synchronized.T0">T0</span></span>)</span><span class="result">: <span class="extype" name="java.lang.AnyRef.synchronized.T0">T0</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.variant.GenotypeRDD#toVariantContextRDD" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="toVariantContextRDD:org.bdgenomics.adam.rdd.variant.VariantContextRDD"></a>
      <a id="toVariantContextRDD:VariantContextRDD"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">toVariantContextRDD</span><span class="result">: <a href="VariantContextRDD.html" class="extype" name="org.bdgenomics.adam.rdd.variant.VariantContextRDD">VariantContextRDD</a></span>
      </span>
      </h4>
      <p class="shortcomment cmt"></p><div class="fullcomment"><div class="comment cmt"></div><dl class="paramcmts block"><dt>returns</dt><dd class="cmt"><p>Returns this GenotypeRDD squared off as a VariantContextRDD.
</p></dd></dl></div>
    </li><li name="org.bdgenomics.adam.rdd.GenomicRDD#transform" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="transform(tFn:org.apache.spark.rdd.RDD[T]=&gt;org.apache.spark.rdd.RDD[T]):U"></a>
      <a id="transform((RDD[Genotype])⇒RDD[Genotype]):GenotypeRDD"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier"></span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">transform</span><span class="params">(<span name="tFn">tFn: (<span class="extype" name="org.apache.spark.rdd.RDD">RDD</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]) ⇒ <span class="extype" name="org.apache.spark.rdd.RDD">RDD</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</span>)</span><span class="result">: <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a></span>
      </span>
      </h4>
      <p class="shortcomment cmt">Applies a function that transforms the underlying RDD into a new RDD.</p><div class="fullcomment"><div class="comment cmt"><p>Applies a function that transforms the underlying RDD into a new RDD.
</p></div><dl class="paramcmts block"><dt class="param">tFn</dt><dd class="cmt"><p>A function that transforms the underlying RDD.</p></dd><dt>returns</dt><dd class="cmt"><p>A new RDD where the RDD of genomic data has been replaced, but the
  metadata (sequence dictionary, and etc) is copied without modification.
</p></dd></dl><dl class="attributes block"> <dt>Definition Classes</dt><dd><a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a></dd></dl></div>
    </li><li name="scala.AnyRef#wait" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="wait():Unit"></a>
      <a id="wait():Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">wait</span><span class="params">()</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd><dt>Annotations</dt><dd>
                <span class="name">@throws</span><span class="args">(<span>
      
      <span class="defval" name="classOf[java.lang.InterruptedException]">...</span>
    </span>)</span>
              
        </dd></dl></div>
    </li><li name="scala.AnyRef#wait" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="wait(x$1:Long,x$2:Int):Unit"></a>
      <a id="wait(Long,Int):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">wait</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.Long">Long</span></span>, <span name="arg1">arg1: <span class="extype" name="scala.Int">Int</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd><dt>Annotations</dt><dd>
                <span class="name">@throws</span><span class="args">(<span>
      
      <span class="defval" name="classOf[java.lang.InterruptedException]">...</span>
    </span>)</span>
              
        </dd></dl></div>
    </li><li name="scala.AnyRef#wait" visbl="pub" data-isabs="false" fullComment="yes" group="Ungrouped">
      <a id="wait(x$1:Long):Unit"></a>
      <a id="wait(Long):Unit"></a>
      <h4 class="signature">
      <span class="modifier_kind">
        <span class="modifier">final </span>
        <span class="kind">def</span>
      </span>
      <span class="symbol">
        <span class="name">wait</span><span class="params">(<span name="arg0">arg0: <span class="extype" name="scala.Long">Long</span></span>)</span><span class="result">: <span class="extype" name="scala.Unit">Unit</span></span>
      </span>
      </h4>
      <div class="fullcomment"><dl class="attributes block"> <dt>Definition Classes</dt><dd>AnyRef</dd><dt>Annotations</dt><dd>
                <span class="name">@throws</span><span class="args">(<span>
      
      <span class="defval" name="classOf[java.lang.InterruptedException]">...</span>
    </span>)</span>
              
        </dd></dl></div>
    </li></ol>
            </div>

        

        
        </div>

        <div id="inheritedMembers">
        <div class="parent" name="scala.Product">
              <h3>Inherited from <span class="extype" name="scala.Product">Product</span></h3>
            </div><div class="parent" name="scala.Equals">
              <h3>Inherited from <span class="extype" name="scala.Equals">Equals</span></h3>
            </div><div class="parent" name="org.bdgenomics.adam.rdd.MultisampleAvroGenomicRDD">
              <h3>Inherited from <a href="../MultisampleAvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.MultisampleAvroGenomicRDD">MultisampleAvroGenomicRDD</a>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>]</h3>
            </div><div class="parent" name="org.bdgenomics.adam.rdd.MultisampleGenomicRDD">
              <h3>Inherited from <a href="../MultisampleGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.MultisampleGenomicRDD">MultisampleGenomicRDD</a>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>]</h3>
            </div><div class="parent" name="org.bdgenomics.adam.rdd.AvroGenomicRDD">
              <h3>Inherited from <a href="../AvroGenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.AvroGenomicRDD">AvroGenomicRDD</a>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>]</h3>
            </div><div class="parent" name="org.bdgenomics.adam.rdd.GenomicRDD">
              <h3>Inherited from <a href="../GenomicRDD.html" class="extype" name="org.bdgenomics.adam.rdd.GenomicRDD">GenomicRDD</a>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>, <a href="" class="extype" name="org.bdgenomics.adam.rdd.variant.GenotypeRDD">GenotypeRDD</a>]</h3>
            </div><div class="parent" name="org.bdgenomics.adam.rdd.ADAMRDDFunctions">
              <h3>Inherited from <span class="extype" name="org.bdgenomics.adam.rdd.ADAMRDDFunctions">ADAMRDDFunctions</span>[<span class="extype" name="org.bdgenomics.formats.avro.Genotype">Genotype</span>]</h3>
            </div><div class="parent" name="org.bdgenomics.utils.misc.Logging">
              <h3>Inherited from <span class="extype" name="org.bdgenomics.utils.misc.Logging">Logging</span></h3>
            </div><div class="parent" name="scala.Serializable">
              <h3>Inherited from <span class="extype" name="scala.Serializable">Serializable</span></h3>
            </div><div class="parent" name="java.io.Serializable">
              <h3>Inherited from <span class="extype" name="java.io.Serializable">Serializable</span></h3>
            </div><div class="parent" name="scala.AnyRef">
              <h3>Inherited from <span class="extype" name="scala.AnyRef">AnyRef</span></h3>
            </div><div class="parent" name="scala.Any">
              <h3>Inherited from <span class="extype" name="scala.Any">Any</span></h3>
            </div>
        
        </div>

        <div id="groupedMembers">
        <div class="group" name="Ungrouped">
              <h3>Ungrouped</h3>
              
            </div>
        </div>

      </div>

      <div id="tooltip"></div>

      <div id="footer">  </div>
      <script defer="defer" type="text/javascript" id="jquery-js" src="../../../../../lib/jquery.js"></script><script defer="defer" type="text/javascript" id="jquery-ui-js" src="../../../../../lib/jquery-ui.js"></script><script defer="defer" type="text/javascript" id="tools-tooltip-js" src="../../../../../lib/tools.tooltip.js"></script><script defer="defer" type="text/javascript" id="template-js" src="../../../../../lib/template.js"></script>
    </body>
      </html>